VmAgent.Core/Interfaces/SystemOperations.cs

```diff
@@ -86,63 +106,82 @@ public void DeleteFile(string filePath)
 
         public void CreateDirectory(string fullPath)
         {
-            Directory.CreateDirectory(fullPath);
+            DirectoryInfo directoryInfo = new DirectoryInfo(fullPath);
+            CreateDirectoryAndParents(directoryInfo);
         }
 
-        // From MSDN: https://msdn.microsoft.com/en-us/library/system.io.directoryinfo.aspx
         public void CopyDirectoryContents(DirectoryInfo source, DirectoryInfo target)
         {
             if (string.Equals(source.FullName, target.FullName, StringComparison.Ordinal))
             {
                 return;
             }
 
-            // Note, if the directory already exists, it doesn't try to recreate it.
-            Directory.CreateDirectory(target.FullName);
+            _fileSystemOperations.Create(target);
+            SetUnixOwnerIfNeeded(target.FullName);
 
             // Copy each file into it's new directory.
-            foreach (FileInfo fi in source.GetFiles())
+            foreach (FileInfo fi in _fileSystemOperations.GetFiles(source))
             {
-                fi.CopyTo(Path.Combine(target.ToString(), fi.Name), true);
+                _fileSystemOperations.CopyTo(fi, Path.Combine(target.ToString(), fi.Name), true);
+                SetUnixOwnerIfNeeded(Path.Combine(target.ToString(), fi.Name));
             }
 
             // Copy each subdirectory using recursion.
-            foreach (DirectoryInfo diSourceSubDir in source.GetDirectories())
+            foreach (DirectoryInfo diSourceSubDir in _fileSystemOperations.GetDirectories(source))
             {
-                DirectoryInfo nextTargetSubDir =
-                    target.CreateSubdirectory(diSourceSubDir.Name);
-                this.CopyDirectoryContents(diSourceSubDir, nextTargetSubDir);
+                //It's ok to create the directory here, the recursion will set the permissions on the next call
+                DirectoryInfo nextTargetSubDir = _fileSystemOperations.CreateSubdirectory(target, diSourceSubDir.Name);
```

Comment:
We don't need this line in that case right? The next call creates the full path anyway (not a big deal either way)

Reply:
It's the easiest way to get the subdirectory path without doing parsing (lazyness)